src/transformers/tokenization_utils_base.py

```diff
@@ -1100,35 +1112,73 @@ def additional_special_tokens(self) -> List[str]:
 
     @bos_token.setter
     def bos_token(self, value):
-        self._bos_token = value
+        self._bos_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @eos_token.setter
     def eos_token(self, value):
-        self._eos_token = value
+        self._eos_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @unk_token.setter
     def unk_token(self, value):
-        self._unk_token = value
+        self._unk_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @sep_token.setter
     def sep_token(self, value):
-        self._sep_token = value
+        self._sep_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @pad_token.setter
     def pad_token(self, value):
-        self._pad_token = value
+        self._pad_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @cls_token.setter
     def cls_token(self, value):
-        self._cls_token = value
+        self._cls_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True)
+            if isinstance(value, str) and value != ""
+            else value
+        )
 
     @mask_token.setter
     def mask_token(self, value):
-        self._mask_token = value
+        self._mask_token = (
+            AddedToken(value, normalized=False, rstrip=True, lstrip=True) if isinstance(value, str) else value
+        )
 
     @additional_special_tokens.setter
     def additional_special_tokens(self, value):
-        self._additional_special_tokens = value
+        # We store the `AddedToken` to allow adding tokens via `tokenizer.add_special_tokens`
+        if value is not None:
+            for token in value:
+                if str(token) not in self.all_special_tokens:
+                    token = (
+                        AddedToken(token, normalized=False, rstrip=True, lstrip=True)
+                        if isinstance(token, str) and token != ""
+                        else token
+                    )
+                    self._additional_special_tokens.append(token)
+
+        else:
+            self._additional_special_tokens = value
```

Comment:
* An early return would make this a bit tidier 
* Safety check

```suggestion
        if value is None:
            self._additional_special_tokens = value
            return
        
        # We store the `AddedToken` to allow adding tokens via `tokenizer.add_special_tokens`
        for token in value:
            if isinstance(token, str) and token != "":
                token = AddedToken(token, normalized=False, rstrip=True, lstrip=True)
            elif not isinstance(token, AddedToken):
                raise ValueError(f"Cannot add instance of type {type(value)} to additional_special_tokens!")
            self._additional_special_tokens.append(token)
```